app/controllers/tokens_controller.rb

```diff
@@ -0,0 +1,69 @@
+class TokensController < Doorkeeper::TokensController
+  include ErrorResponses
+
+  def create
+    authenticate_with_facebook if signing_in_with_facebook?
+    authenticate_with_credentials if singing_in_regularly?
+  rescue Doorkeeper::Errors::DoorkeeperError,
+         Doorkeeper::Errors::InvalidGrantReuse,
+         Doorkeeper::OAuth::Error,
+         Koala::Facebook::AuthenticationError,
+         ActiveRecord::RecordNotFound => e
+    render_error e
+  end
+
+  private
+
+    def signing_in_with_facebook?
+      params[:username] == "facebook"
```

Comment:
This was taken from code-corps-api and fieldthebern before that one. I believe you were the one implementing it and didn't manage to find a better way. I do agree it's far from ideal, so I'll create an issue to revisit.